source/material_model/rheology/elasticity.cc

```diff
@@ -0,0 +1,381 @@
+                const Tensor<2,dim> rotation = 0.5 * ( old_velocity_gradients[i] - transpose(old_velocity_gradients[i]) );
+
+                // Average viscoelastic viscosity
+                const double average_viscoelastic_viscosity = out.viscosities[i];
+
+                // Calculate the current (new) viscoelastic stress, which is a function of the material
+                // properties (viscoelastic viscosity, shear modulus), elastic time step size, strain rate,
+                // vorticity and prior (inherited) viscoelastic stresses (see equation 29 in Moresi et al.,
+                // 2003, J. Comp. Phys.)
+                SymmetricTensor<2,dim> stress_new = ( 2. * average_viscoelastic_viscosity * deviator(in.strain_rate[i]) ) +
+                                                    ( ( average_viscoelastic_viscosity / ( average_elastic_shear_moduli[i] * dte ) ) * stress_old ) +
+                                                    ( ( average_viscoelastic_viscosity / average_elastic_shear_moduli[i] ) *
+                                                      ( symmetrize(rotation * Tensor<2,dim>(stress_old) ) - symmetrize(Tensor<2,dim>(stress_old) * rotation) ) );
+
+                // Stress averaging scheme to account for difference betweed fixed elastic time step
+                // and numerical time step (see equation 32 in Moresi et al., 2003, J. Comp. Phys.)
+                if (use_fixed_elastic_time_step == true && use_stress_averaging == true)
+                  {
+                    stress_new = ( ( 1. - ( dt / dte ) ) * stress_old ) + ( ( dt / dte ) * stress_new ) ;
+                  }
+
+                // Fill reaction terms
+                for (unsigned int j = 0; j < SymmetricTensor<2,dim>::n_independent_components ; ++j)
+                  out.reaction_terms[i][j] = -stress_old[SymmetricTensor<2,dim>::unrolled_to_component_indices(j)]
+                                             + stress_new[SymmetricTensor<2,dim>::unrolled_to_component_indices(j)];
+
+              }
+          }
+      }
+
+
+
+      template <int dim>
+      double
+      Elasticity<dim>::elastic_timestep () const
+      {
+        // The elastic time step (dte) is equal to the numerical time step if the time step number
+        // is greater than 0 and the parameter 'use_fixed_elastic_time_step' is set to false.
+        // On the first (0) time step the elastic time step is always equal to the value
+        // specified in 'fixed_elastic_time_step', which is also used in all subsequent time
+        // steps if 'use_fixed_elastic_time_step' is set to true.
+        //
+        // We also use this parameter when we are still *before* the first time step,
+        // i.e., if the time step number is numbers::invalid_unsigned_int.
+        const double dte = ( ( this->get_timestep_number() > 0 &&
+                               this->get_timestep_number() != numbers::invalid_unsigned_int &&
```

Comment:
This should be equivalent to 

```suggestion
                               this->simulator_is_past_initialization() &&
```